src/cvar.ml

```diff
@@ -26,15 +26,38 @@ struct
     let of_index v = Var v
   end
 
+  let scratch = Field.of_int 0
+
   let eval context t0 =
-    let res = ref Field.zero in
+    let open Field in
+    let res = of_int 0 in
+    let can_mutate_scale = ref false in
```

Comment:
Is this actually safe? Like if the scalar is Field.one, wouldn't this mutate Field.one?

Reply:
`scale` is an internal variable only set within this function, and `can_mutate_scale` is updated every time its status changes. So
* we call `go one t0` (ie. `scale = one`) from the outer `eval` with `can_mutate_scale = false` (to avoid an extra field element allocation on every call to `eval`)
* we set `can_mutate_scale = false` for the first branch of an `Add`, resetting it to its original value for the second, to make sure that `scale` is preserved for that second branch
* we set `can_mutate_scale = true` when we see a `Scale` with `can_mutate_scale = false`, since the `Field.mul` we do there gives us a fresh field element (and reset it after, because we know the parent context has an immutable `scale`).